core/src/toga/hardware/camera.py

```diff
@@ -0,0 +1,153 @@
+from __future__ import annotations
+
+from typing import TYPE_CHECKING
+
+from toga.constants import FlashMode
+from toga.handlers import AsyncResult
+from toga.platform import get_platform_factory
+
+if TYPE_CHECKING:
+    from toga.app import App
+
+
+class PermissionResult(AsyncResult):
+    RESULT_TYPE = "permission"
+
+
+class PhotoResult(AsyncResult):
+    RESULT_TYPE = "photo"
+
+
+class CameraDevice:
+    def __init__(self, impl):
+        self._impl = impl
+
+    @property
+    def id(self) -> str:
+        """A unique identifier for the device"""
+        return self._impl.id()
+
+    @property
+    def name(self) -> str:
+        """A human-readable name for the device"""
+        return self._impl.name()
+
+    @property
+    def has_flash(self) -> bool:
+        """Does the device have a flash?"""
+        return self._impl.has_flash()
+
+    def __eq__(self, other) -> bool:
+        return self.id == other.id
+
+    def __repr__(self) -> str:
+        return f"<CameraDevice id={self.id} {self.name!r}>"
+
+    def __str__(self) -> str:
+        return self.name
```

Comment:
Why have both `str` and `repr`? For classes like this, `repr` is normally good enough on its own, and avoids the confusion of the `str` return value not making it evident what the object's type is. If the user really wants a string, they can call `id` or `name` directly.

Reply:
Why *ever* have `repr`? It's for debugging. In this case, I was hitting a debugging case where the default repo was giving me a memory address, but that wasn't helping me debug what was going on, because `__eq__` allows two different instances to be equal.

Reply:
My suggestion was to keep `__repr__` and remove `__str__`; sorry for not making that clear.

Reply:
`__str__` is actually useful (and publicly visible) functionality. It means the camera selection pulldown on the camera popup window on macOS can use a "list of CameraDevice" as it's data source, and the selected item is literally Camera instances, rendered using the `__str__` method.

Reply:
OK.